compiler/defer.go

```diff
@@ -36,6 +63,123 @@ func (b *builder) deferInitFunc() {
 	deferType := llvm.PointerType(b.getLLVMRuntimeType("_defer"), 0)
 	b.deferPtr = b.CreateAlloca(deferType, "deferPtr")
 	b.CreateStore(llvm.ConstPointerNull(deferType), b.deferPtr)
+
+	if b.hasDeferFrame() {
+		// Set up the defer frame with the current stack pointer.
+		// This assumes that the stack pointer doesn't move outside of the
+		// function prologue/epilogue (an invariant maintained by TinyGo but
+		// possibly broken by the C alloca function).
+		// The frame pointer is _not_ saved, because it is marked as clobbered
+		// in the setjmp-like inline assembly.
+		deferFrameType := b.getLLVMType(b.program.ImportedPackage("internal/task").Members["DeferFrame"].Type())
+		b.deferFrame = b.CreateAlloca(deferFrameType, "deferframe.buf")
+		stackPointer := b.readStackPointer()
+		b.createRuntimeCall("setupDeferFrame", []llvm.Value{b.deferFrame, stackPointer}, "")
+
+		// Create the landing pad block, which is where control transfers after
+		// a panic.
+		b.landingpad = llvm.AddBasicBlock(b.llvmFn, "lpad")
+	}
+}
+
+// createLandingPad fills in the landing pad block. This block runs the deferred
+// functions and returns (by jumping to the recover block). If the function is
+// still panicking after the defers are run, the panic will be re-raised in
+// destroyDeferFrame.
+func (b *builder) createLandingPad() {
+	b.SetInsertPointAtEnd(b.landingpad)
+
+	// Add debug info, if needed.
+	// The location used is the closing bracket of the function.
+	if b.Debug {
+		pos := b.program.Fset.Position(b.fn.Syntax().End())
+		b.SetCurrentDebugLocation(uint(pos.Line), uint(pos.Column), b.difunc, llvm.Metadata{})
+	}
+
+	b.createRunDefers()
+
+	// Continue at the 'recover' block, which returns to the parent in an
+	// appropriate way.
+	b.CreateBr(b.blockEntries[b.fn.Recover])
+}
+
+// createInvokeCheckpoint saves the function state at the given point, to
+// continue at the landing pad if a panic happened. This is implemented using a
+// setjmp-like construct.
+func (b *builder) createInvokeCheckpoint() {
+	// Construct inline assembly equivalents of setjmp.
+	// The assembly works as follows:
+	//   * All registers (both callee-saved and caller saved) are clobbered
+	//     after the inline assembly returns.
+	//   * The assembly stores the address just past the end of the assembly
+	//     into the jump buffer.
+	//   * The return value (eax, rax, r0, etc) is set to zero in the inline
+	//     assembly but set to an unspecified non-zero value when jumping using
+	//     a longjmp.
+	asmType := llvm.FunctionType(b.uintptrType, []llvm.Type{b.deferFrame.Type()}, false)
```

Comment:
Are we sure that this is safe? As is, I think the compiler is still permitted to assume that this returns once.

Reply:
I have tried several alternatives, and this was the most reliable one so far.

I've tried:

1. `setjmp` at the beginning of a function. LLVM is smart enough to optimize away all `defer` instructions later after a `longjmp`, so this is not usable.
2. Using exception handling instructions together with `blockaddress`. Basically, act as if we're unwinding as usual with `invoke` and `landingpad`, but instead of actually unwinding, return back to the function by capturing the stack pointer at function entry and doing `longjmp` by jumping back to the basic block with the `landingpad` instruction with a captured `blockaddress`. Doesn't work because LLVM moves the `landingpad` instruction to a different basic block after which the entire system falls down.

The only other alternative is using `setjmp` directly, which would be much more expensive than it is already (because we need to do it before every function call that might panic).
I _think_ this is safe because this is what LLVM says about `returns_twice`:

> This attribute indicates that this function can return twice. The C `setjmp` is an example of such a function. The compiler disables some optimizations (like tail calls) in the caller of these functions.

Tail calls are not relevant to inline assembly. I'm not 100% convinced, but it appears to be mostly safe.

... actually, this gives me an idea. It might be possible to change this scheme to do the `setjmp` after every defer instruction, instead of before every call. Not entirely sure whether that's safe though.

Reply:
Does gollvm support defer/recover? How does it convince llvm to do the Right Thing? ?

Reply:
It looks like gollvm uses full C++ exceptions. I don't think that is practical here.

Reply:
I might start using C++ exceptions on some platforms (Windows, Linux, MacOS) but wanted to start with something simple and mostly portable.

Reply:
@niaow I'm afraid you are correct. It looks like the remaining issue is a tail call on RISC-V, which would probably have been prevented with a `returns_twice`. Now thinking what the best solution to this is...

Reply:
Well that was easy. Apparently it's possible to add `returns_twice` as a call site attribute, even to inline assembly.

```diff
        }
        asm := llvm.InlineAsm(asmType, asmString, constraints, false, false, 0)
        result := b.CreateCall(asm, []llvm.Value{b.deferFrame}, "setjmp")
+       result.AddCallSiteAttribute(-1, b.ctx.CreateEnumAttribute(llvm.AttributeKindID("returns_twice"), 0))
        isZero := b.CreateICmp(llvm.IntEQ, result, llvm.ConstInt(b.uintptrType, 0, false), "setjmp.result")
        continueBB := b.insertBasicBlock("")
        b.CreateCondBr(isZero, continueBB, b.landingpad)
```

Reply:
> (Sidenote: I might investigate using the callbr instruction instead, which might be slightly less expensive by avoiding a compare and branch).

It seems like that might avoid more than just a compare and branch. Would something like this work:
```
callbr void asm returns_twice "", "=~{rax},..."()
            to label %fallthrough [label %indirect]
```

Seems a bit cursed though.

Reply:
I'm not sure what the code does there? I see no assembly.
My biggest hope is that I can manage to avoid clobbering all registers for every function call, and instead can hopefully do that in the `%lpad` block. That will certainly improve code quality.